### PR TITLE
[v2.0] Update CMake code to assume the C++11 standard

### PR DESCRIPTION
**Pull request by krivenko**
* Require CMake 3.1.0 or newer so that we can use `set(CMAKE_CXX_STANDARD 11)`.
* Remove macro `set_cxx11()` and recycle some of its parts in another new macro, `compiler_quirks()`.
* Remove `POMEROL_CXX11` from `first_include.h`.
* Try to use preinstalled gftools first, then download them from GitHub.
* Update programs. In particular, rename `anderson.cxx11.cpp` -> `anderson_gftools.cpp`.
* Bump pomerol version to 2.0.
* Update README.md.

Question: Do we still want/need to maintain both `anderson.cpp` and `anderson_gftools.cpp`?

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -77,8 +78,9 @@ set (pomerol_sources
     pomerol/EnsembleAverage
 )
 
-configure_file(${CMAKE_SOURCE_DIR}/include/pomerol.h.in ${CMAKE_BINARY_DIR}/include/pomerol.h)
```
Comment (aeantipov) [inline CMakeLists.txt:80]: Igor, do you remember the reason of splitting `first_include.h.in` and `pomerol.h.in` ?

File: cmake/CommonDefs.cmake [hunk 2]
```diff
@@ -80,11 +72,11 @@ else()
...
 
-# 
+#
 # Dependencies
 #
 
-# ALPSCore - brings MPI, HDF5, boost 
+# ALPSCore - brings MPI, HDF5, boost
 macro(add_alpscore)
 find_package(ALPSCore REQUIRED COMPONENTS hdf5 accumulators mc params)
```
Comment (aeantipov) [inline cmake/CommonDefs.cmake:81]: mc looks redundant here. I am not sure if alpscore::accumulators still exist. Do we actually need the dependency at all? Looking at the amount of code we use (https://github.com/aeantipov/pomerol/search?q=alpscore) - it's an absolute minimum. Feel free to address this in a separate PR if needed.

File: cmake/CommonDefs.cmake [hunk 3]
```diff
@@ -254,7 +235,7 @@ macro(add_opendf_module)
         else(DEFINED BOLD_HYB_GLOBAL_BUILD)
             string(REGEX REPLACE "^opendf_" "" pkgcomp_ ${pkg_})
             find_package(opendf QUIET COMPONENTS ${pkgcomp_})
-            if (opendf_${pkgcomp_}_FOUND)
```
Comment (aeantipov) [inline cmake/CommonDefs.cmake:257]: this probably should be pomerol, not opendf :)

**Review by aeantipov — APPROVED** [review]

**Comment by aeantipov** [comment]
Re: `anderson` and `anderson_gftools` - I am not sure. The difference is that anderson.cpp can be compiled without dependency, whereas anderson_gftools needs both gftools and alpscore (for hdf5). We depend on headers of gftools but not on the linked code, so in principle `anderson_gftools` is optional and current structure supports that. On the other hand, supporting less code is always a good idea. What do you think?

File: CMakeLists.txt (hunk 1 above)
Comment (krivenko) [reply CMakeLists.txt:80]: At this point I can only guess, and my guess is that including `first_include.h` gives you access to all those configuration macros without including the rest of pomerol headers.

I am going to remove this `configure_file()` as there is nothing to configure in `pomerol.h.in`.

File: cmake/CommonDefs.cmake (hunk 2 above)
Comment (krivenko) [reply cmake/CommonDefs.cmake:81]: > I am not sure if alpscore::accumulators still exist.

FYI, they do exist in the deprecated status.
New codes are meant to depend on [alea](https://github.com/ALPSCore/ALPSCore/tree/master/alea) instead (see also [this preprint](https://arxiv.org/abs/1811.08331), which has been under peer review for almost 2 years now).

> Do we actually need the dependency at all?

No, we do not. `add_alpscore()` is defined here but never called.
`https://github.com/aeantipov/pomerol/search?q=alpscore` finds only comments.

I'd rather remove this macro altogether and keep only that mention of ALPSCore in `README.md`.

**Comment by krivenko** [comment]
> The difference is that anderson.cpp can be compiled without much dependencies, whereas anderson_gftools needs both gftools and alpscore (for hdf5).

As it stands now, `anderson_gftools` does not need ALPSCore, and gftools will always be either detected or downloaded when `-DProgs=ON`.
But sure, I see no problem in keeping both.

File: cmake/CommonDefs.cmake (hunk 3 above)
Comment (krivenko) [reply cmake/CommonDefs.cmake:257]: Um, why does this project need a macro called `add_opendf_module()`, which also refers to `BOLD_HYB_GLOBAL_BUILD`? :)
It is never called in any CMake script.

**Comment by aeantipov** [comment]
> > The difference is that anderson.cpp can be compiled without much dependencies, whereas anderson_gftools needs both gftools and alpscore (for hdf5).
> 
> As it stands now, `anderson_gftools` does not need ALPSCore, and gftools will always be either detected or downloaded when `-DProgs=ON`.
> But sure, I see no problem in keeping both.

In this case I suggest we remove anderson.cpp. Since there are no actual extra dependencies - no need to use both.

File: cmake/CommonDefs.cmake (hunk 3 above)
Comment (aeantipov) [reply cmake/CommonDefs.cmake:257]: History of copy-pasting between codes be like... :)

Comment (aeantipov) [reply cmake/CommonDefs.cmake:257]: Let's add a TODO to clean this if you don't want to address it now.

File: cmake/CommonDefs.cmake (hunk 2 above)
Comment (aeantipov) [reply cmake/CommonDefs.cmake:81]: Let's remove it.

**Comment by krivenko** [comment]
@aeantipov I am done with this round of cleaning. Let me know if you are fine with merging.

**Comment by aeantipov** [comment]
Yup. Feel free to merge whenever all conversations are resolved. Thanks!

**Pull request merged**